Modules/_json.c

```diff
@@ -1482,7 +1482,7 @@ encoder_listencode_obj(PyEncoderObject *s, _PyAccu *acc,
         return _steal_accumulate(acc, encoded);
     }
     else if (PyLong_Check(obj)) {
-        PyObject *encoded = PyLong_Type.tp_str(obj);
+        PyObject *encoded = PyLong_Type.tp_repr(obj);
```

Comment:
What happens to user-side code that did this?

Reply:
Some history.

Originally, we called `str()` for integers and `repr()` for floats, because `repr()` for `long` in Python 2 contained the "L" suffix, and `str()` for floats was shorter than `repr()`, with the loss of information. We needed just decimal representation without losses. Later we discovered that this produces invalid JSON for int and float subclasses wish overridden `__str__` or `__repr__`, so we started to call `int.__str__` and `float.__repr__` directly. Now, after removing `int.__str__`, `int.__str__` is equal to `object.__str__` which calls `repr()`. But we need to call `int.__repr__` instead of the `__repr__` overridden in int subclass.

Reply:
I specifically meant code calling `PyLong_Type.tp_str()`, like in the case above. Cython does this in its f-string formatting code. I can obviously change it to use `PyLong_Type.tp_repr()`, but only in a future release, and it would also only apply to Py3, since Py2 gives a different result than `tp_str()`. Sure, I could also just call `PyObject_Str()` (or `PyObject_Repr()` in Py3, to avoid yet another indirection), but that seems a bit wasteful, given that we already know at this point that it will end up calling `PyLong_Type.tp_str()`.

I just think it makes a difference whether this change is made in Python code, or whether it changes the `PyTypeObject` struct in C. I understand the change and the history behind it, and I think it's better if, for example, int subclasses don't need to override both `__str__` and `__repr__` in Python. I just think it's worth noting that this might be a more breaking change than it looks at first.

Reply:
Ok, I looked up the inheritance mechanism. If `tp_str` is empty, `PyType_Ready()` sets it to `object_str`, which delegates to `tp_repr`. So this change doesn't actually break anything.